packages/xsnap/src/avaAssertXS.js

```diff
@@ -0,0 +1,315 @@
+/* eslint-disable no-await-in-loop */
+// @ts-check
+
+/** global print */
+
+const { freeze, keys } = Object;
+
+// ack Paul Roub Aug 2014
+// https://stackoverflow.com/a/25456134/7963
+/** @type {(x: unknown, y: unknown) => boolean } */
+function deepEqual(x, y) {
+  if (x === y) {
+    return true;
+  }
+  if (
+    typeof x === 'object' &&
+    x != null &&
+    typeof y === 'object' &&
+    y != null
+  ) {
+    if (keys(x).length !== keys(y).length) {
+      const detail = JSON.stringify({
+        actual: {
+          length: keys(x).length,
+          keys: keys(x),
+        },
+        expected: {
+          length: keys(y).length,
+          keys: keys(y),
+        },
+      });
+      throw new Error(`Object keys length: ${detail}`);
+    }
+
+    for (const prop in x) {
+      // eslint-disable-next-line no-prototype-builtins
+      if (y.hasOwnProperty(prop)) {
+        if (!deepEqual(x[prop], y[prop])) {
+          return false;
+        }
+      } else {
+        throw new Error(`missing property ${prop}`);
+      }
+    }
+
+    return true;
+  }
+  const detail = JSON.stringify({
+    actual: { type: typeof x, value: x },
+    expected: { type: typeof y, value: y },
+  });
+  throw new Error(detail);
```

Comment:
Does this error detail structure come from Ava?

Reply:
No.